### PR TITLE
Add GenesisHash to ActionEvaluator

### PR DESCRIPTION
**Pull request by hunkim98**
Added GenesisHash property to ActionEvaluator class

Resolves issue #1972

### CODE REVIEW

File: Libplanet.Tests/Action/ActionEvaluatorTest.cs [hunk 1]
```diff
@@ -581,7 +583,7 @@ public void EvaluateTx()
             var actionEvaluator = new ActionEvaluator<DumbAction>(
                 policyBlockAction: null,
                 blockChainStates: NullChainStates<DumbAction>.Instance,
-                trieGetter: null);
+                trieGetter: null, genesisHash: null);
```
Comment (dahlia) [inline Libplanet.Tests/Action/ActionEvaluatorTest.cs:586]: For consistency, please give a line to each argument.

```suggestion
                trieGetter: null,
                genesisHash: null);
```

File: Libplanet.Tests/Action/ActionEvaluatorTest.cs [hunk 2]
```diff
@@ -708,7 +710,7 @@ public void EvaluateTxResultThrowingException()
             var actionEvaluator = new ActionEvaluator<ThrowException>(
                 policyBlockAction: null,
                 blockChainStates: NullChainStates<ThrowException>.Instance,
-                trieGetter: null);
+                trieGetter: null, genesisHash: null);
```
Comment (dahlia) [inline Libplanet.Tests/Action/ActionEvaluatorTest.cs:713]: For consistency, please give a line to each argument.

```suggestion
                trieGetter: null,
                genesisHash: null);
```

File: Libplanet/Action/ActionEvaluator.cs [hunk 3]
```diff
@@ -40,15 +41,19 @@ public class ActionEvaluator<T>
         /// the states for a provided <see cref="Address"/>.</param>
         /// <param name="trieGetter">The function to retrieve a trie for
         /// a provided <see cref="BlockHash"/>.</param>
+        /// <param name="genesisHash"> A <see cref="HashDigest{SHA256}"/> value of the genesis block
```
Comment (dahlia) [inline Libplanet/Action/ActionEvaluator.cs:44]: It doesn't seem a `HashDigest<SHA256>`:

```suggestion
        /// <param name="genesisHash">A <see cref="BlockHash"/> value of the genesis block
```

**Review by dahlia — CHANGES_REQUESTED** [review]
Thanks for your first contribution!  However, this patch does not completely address #1972, as `IActionContext.GenesisHash` is still unimplemented.

File: Libplanet.Tests/Action/ActionEvaluatorTest.cs (hunk 1 above)
Comment (hunkim98) [reply Libplanet.Tests/Action/ActionEvaluatorTest.cs:586]: Sure! I will look into it again

**Comment by hunkim98** [comment]
I have applied the suggestions advised by @dahlia and properly added the genesis hash to IActionContext. I also have changed a few test codes to remedy the changes applied to IActionContext.

File: Libplanet/Action/IActionContext.cs [hunk 4]
```diff
@@ -11,6 +11,12 @@ namespace Libplanet.Action
     /// </summary>
     public interface IActionContext
     {
+        /// <summary>
+        ///  A <see cref="BlockHash"/> value of the genesis block
```
Comment (dahlia) [inline Libplanet/Action/IActionContext.cs:15]:
```suggestion
        ///  The genesis block's hash.
```

File: Libplanet/Action/IActionContext.cs [hunk 5]
```diff
@@ -11,6 +11,12 @@ namespace Libplanet.Action
...
+        /// </summary>
+        [Pure]
+        BlockHash? GenesisHash { get; }
```
Comment (dahlia) [inline Libplanet/Action/IActionContext.cs:18]: This backward-incompatible change should be recorded in the *CHANGES.md* file.

File: Libplanet/Action/ActionEvaluator.cs [hunk 6]
```diff
@@ -40,15 +41,19 @@ public class ActionEvaluator<T>
...
         /// <param name="trieGetter">The function to retrieve a trie for
         /// a provided <see cref="BlockHash"/>.</param>
+        /// <param name="genesisHash"> A <see cref="BlockHash"/> value of the genesis block
+        /// </param>
         public ActionEvaluator(
             IAction? policyBlockAction,
             IBlockChainStates<T> blockChainStates,
-            Func<BlockHash, ITrie>? trieGetter)
+            Func<BlockHash, ITrie>? trieGetter,
+            BlockHash? genesisHash)
```
Comment (dahlia) [inline Libplanet/Action/ActionEvaluator.cs:50]: This backward-incompatible change should be recorded in the *CHANGES.md* file.

**Review by dahlia — COMMENTED** [review]
Could we have a substantial unit test to ensure if `IActionContext.GenesisHash` is accessible inside actions and its value is correct?

File: Libplanet/Action/ActionEvaluator.cs [hunk 7]
```diff
@@ -184,6 +190,7 @@ internal static IImmutableSet<Address> GetUpdatedAddresses(Transaction<T> tx)
         /// Executes <see cref="IAction"/>s in <paramref name="actions"/>.  All other evaluation
         /// calls resolve to this method.
         /// </summary>
+        /// <param name="genesisHash"> A <see cref="BlockHash"/> value of the genesis block </param>
```
Comment (riemannulus) [inline Libplanet/Action/ActionEvaluator.cs:193]: Our docstring lint is required to line end with a period.
```suggestion
        /// <param name="genesisHash"> A <see cref="BlockHash"/> value of the genesis block. </param>
```

**Review by riemannulus — CHANGES_REQUESTED** [review]

**Comment by hunkim98** [comment]
> Could we have a substantial unit test to ensure if `IActionContext.GenesisHash` is accessible inside actions and its value is correct?

I'll try it out! In which folder or file should I implement the unit test? (Should I create a new file for it? - I'm not sure how to create a unit test for a C# library - any references would be welcom!)

**Comment by dahlia** [comment]
> In which folder or file should I implement the unit test? (Should I create a new file for it? - I'm not sure how to create a unit test for a C# library - any references would be welcome!)

@hunkim98 [This section in *CONTRIBUTING.md*][1] will help you to hang of it briefly.  IMHO you don't have to create a new file; instead, adding a new method to the existing *ActionEvaluatorTest* class would be enough.

[1]: https://github.com/planetarium/libplanet/blob/main/CONTRIBUTING.md#tests--

File: CHANGES.md [hunk 8]
```diff
@@ -48,6 +51,8 @@ To be released.
 [#2159]: https://github.com/planetarium/libplanet/pull/2159
 [#2162]: https://github.com/planetarium/libplanet/issues/2162
 [#2171]: https://github.com/planetarium/libplanet/pull/2171
+[#1972]: https://github.com/planetarium/libplanet/issues/1972
```
Comment (longfin) [inline CHANGES.md:54]: Keeping sorting would be good. 😅

File: CHANGES.md [hunk 9]
```diff
@@ -40,6 +42,7 @@ To be released.
 
 ### CLI tools
 
+
```
Comment (OnedgeLee) [inline CHANGES.md:45]:
```suggestion
```

**Review by OnedgeLee — CHANGES_REQUESTED** [review]
LGTM. After commit suggestion and squashing commit, I think it can be merged.

File: Libplanet.Tests/Action/ActionEvaluatorTest.cs [hunk 10]
```diff
@@ -1071,6 +1077,29 @@ public void EvaluatePolicyBlockAction()
             return (addresses, txs);
         }
 
+        [Fact]
+        private async void CheckGenesisHashInAction()
```
Comment (longfin) [inline Libplanet.Tests/Action/ActionEvaluatorTest.cs:1081]:
```suggestion
        private async Task CheckGenesisHashInAction()
```

Generally, `async Task` is more affordable than `async void` [due to controllability](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void). 
Fortunately, xUnit seems to support `async void` correctly, but I still prefer `async Task` for portability 😅

**Review by longfin — APPROVED** [review]
@hunkim98 Thanks for your contribution! 🎉 I left two minor comments but I guess they aren't blockers 😄

**Review by longfin — APPROVED** [review]
LGTM

**Review by OnedgeLee — DISMISSED** [review]
LGTM

File: CHANGES.md [hunk 11]
```diff
@@ -26,6 +26,8 @@ To be released.
    - Added `BlockChain<T>.FindNextHashes(BlockLocator, BlockHash?, int)` method.
    - Added `BlockChain<T>.Fork(BlockHash, bool)` method.
    - Added `BlockChain<T>.GetBlockLocator(int)` method.
+ -  Added Genesis Hash property in `IActionContext`. [[#1972], [#2179]]
+   -  Added Genesis Hash property in `ActionEvaluator`.
```
Comment (dahlia) [inline CHANGES.md:30]: Some stylistic adjustments:

```suggestion
 -  Added `IActionContext.GenesisHash` property.  [[#1972], [#2179]]
 -  Added `ActionEvaluator<T>.GenesisHash` property.  [[#1972], [#2179]]
```

**Review by OnedgeLee — APPROVED** [review]

**Review by dahlia — APPROVED** [review]

**Review by longfin — APPROVED** [review]

**Pull request merged**